wgkey/key.go

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package wgkey contains types and helpers for WireGuard keys.
+package wgkey
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/subtle"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strings"
+
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/curve25519"
+)
+
+// Size is the number of bytes in a curve25519 key.
+const Size = 32
+
+// A Key is a curve25519 key.
+// It is used by WireGuard to represent public and preshared keys.
```

Comment:
Clarify that Tailscale never uses preshared keys so this is always a public key? I think that's true?

Reply:
We don't use preshared keys in production, but some test code constructs preshared keys. I'll change that and make appropriate changes in a follow-up because I don't want this large change to rot.